src/HttpOverStream.Client/DialMessageHandler.cs

```diff
@@ -62,42 +63,76 @@ protected override Task<Stream> CreateContentReadStreamAsync()
         protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
             ValidateAndNormalizeRequest(request);
-            var stream = await _dial.DialAsync(request, cancellationToken).ConfigureAwait(false);
-
-            request.Properties.Add(UnderlyingStreamProperty, stream);
-            await stream.WriteMethodAndHeadersAsync(request, cancellationToken).ConfigureAwait(false);
-            if (request.Content != null)
+            Stream stream = null;
+            try
             {
-                await request.Content.CopyToAsync(stream).ConfigureAwait(false);
-            }
-            await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+                Debug.WriteLine("Client: About to connect");
+                stream = await _dial.DialAsync(request, cancellationToken).ConfigureAwait(false);
 
-            var responseContent = new DialResponseContent();
-            var response = new HttpResponseMessage { RequestMessage = request, Content = responseContent };
-            string statusLine = await stream.ReadLineAsync(cancellationToken).ConfigureAwait(false);
-            ParseStatusLine(response, statusLine);
-            for (; ; )
-            {
-                var line = await stream.ReadLineAsync(cancellationToken).ConfigureAwait(false);
-                if (line.Length == 0)
+                Debug.WriteLine("Client: Connected");
+                request.Properties.Add(UnderlyingStreamProperty, stream);
+
+                Debug.WriteLine("Client: Writing request");
+                await stream.WriteMethodAndHeadersAsync(request, cancellationToken).ConfigureAwait(false);
+
+                // as soon as headers are sent, we should begin reading the response, and send the request body concurrently
+                // This is because if the server 404s nothing will ever read the response and it'll hang waiting
+                // for someone to read it
+                var writeContentCTS = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                var writeContentToken = writeContentCTS.Token;
+                var writeContentTask = Task.Run(async () => // Cancel this task if server response detected
+                    {
+                        if (request.Content != null)
+                        {
+                            Debug.WriteLine("Client: Writing request request.Content.CopyToAsync");
+                            await request.Content.CopyToAsync(stream).ConfigureAwait(false);

```

Comment:
missing the writeContentToken here ? (I know NamedPipeStream does not takes it into account, but other streams might)

Reply:
CopyToAsync doesnt have an overload with a cancellationToken =[

Reply:
Ugh :(